torch/_dynamo/compiled_autograd.py

```diff
@@ -183,3 +196,108 @@ def disable():
     prior = torch._C._dynamo.compiled_autograd.set_autograd_compiler(None)
     yield
     torch._C._dynamo.compiled_autograd.set_autograd_compiler(prior)
+
+
+# Invoke backward hook is a higher order op meant for both invoking a bound hook,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the hook as needed. This, in turn, means we can support hooks in backward with complex python
+# state mutation. If we were to not do this, the hooks would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def _invoke_in_backward(*args, fn, reenter):
```

Comment:
I think mediating whether or not the higher order op stays a hop, or if it evaporates into an actual call function, using the "reenter" argument is an unintuitive way of structuring it.

What would make more sense to me is to have a higher order op that *always* preserves itself in proxy tensor tracing, but that Dynamo knows to trace into the real implementation of. Because the point of the operator is to let you say "Hey, I want to call this function that is not traceable, so I need you to stick it in the graph directly". In other words, it's basically a way of making function calls into "leaf modules" (in traditional FX parlance). I wouldn't call it "allow_in_graph" (even though that's basically what it does) because allow_in_graph implies that it is safely traceable, but this is never traceable.

Reply:
Yeah, I like this plan